src/Struct.php

```diff
@@ -0,0 +1,223 @@
+    {
+        $props = $this->getPropTypes();
+
+        $this->__data = [];
+        foreach ($props as $key => $type) {
+            $this->__data[$key] = isset($data[$key])
+                ? $type->cast($data[$key])
+                : $type->getDefault();
+
+            unset($data[$key]);
+        }
+
+        if (!empty($data)) {
+            $class = static::class;
+            $props = implode(', ', array_keys($data));
+            throw new LogicException("Struct {$class} does not have the following properties: {$props}");
+        }
+    }
+
+    /**
+     * Retrieves the value for a property.
+     *
+     * @since [*next-version*]
+     *
+     * @param string $name The name of the property.
+     *
+     * @return mixed|null The value of the property.
+     */
+    public function __get($name)
+    {
+        $props = $this->getPropTypes();
+
+        if (!array_key_exists($name, $props)) {
+            throw static::createUndefinedPropException($name);
+        }
+
+        return $this->__data[$name];
+    }
+
+    /**
+     * Should not be used. Only exists to warn developers of misuse.
+     *
+     * @since [*next-version*]
+     *
+     * @param string $name  The name of the property.
+     * @param mixed  $value The value being assigned to the property.
+     *
+     * @throws Exception Always throws.
+     */
+    public function __set($name, $value)
+    {
+        throw new LogicException('Cannot set property values. Use with() instead');
+    }
+
+    /**
+     * Creates a copy of the struct instance with some modified properties.
+     *
+     * @since [*next-version*]
+     *
+     * @param array $changes An associative array that maps property names to their new values.
+     *
+     * @return Struct The copied struct with the applied changes. If the $changes parameter is empty, no copy is
+     *                performed and the same instance is returned.
+     */
+    public function with(array $changes)
+    {
+        if (empty($changes)) {
+            return $this;
+        }
+
+        $clone = clone $this;
+
+        $props = $this->getPropTypes();
+        foreach ($changes as $key => $value) {
+            if (!isset($props[$key])) {
+                throw static::createUndefinedPropException($key);
+            }
+
+            $clone->__data[$key] = $props[$key]->cast($value);
+        }
+
+        return $clone;
+    }
+
+    /**
+     * Checks if two structs have equal state.
+     *
+     * This method compares the two structs using loose value equivalence (==).
+     *
+     * @since [*next-version*]
+     *
+     * @param Struct $struct The struct to compare to.
+     *
+     * @return bool True if the two structs have equal states, false otherwise.
+     */
+    public function isEqualTo(Struct $struct)
+    {
+        $thisPropsTypes = $this->getPropTypes();
+        $otherPropTypes = $struct->getPropTypes();
+
+        return $this->__data == $struct->__data && $thisPropsTypes == $otherPropTypes;
```

Comment:
Does loose array comparison actually do this?? :O

Reply:
Yeah. I was surprised too.